src/main/java/org/eclipse/yasson/internal/model/ClassModel.java

```diff
@@ -183,6 +184,15 @@ public void setProperties(List<PropertyModel> parsedProperties) {
      * @return default constructor
      */
     public Constructor<?> getDefaultConstructor() {
-        return defaultConstructor;
+        /*
+        Lazy-loads the default constructor to avoid Java 9+ "Illegal reflective access" warnings where possible.
+        Example: Deserialization into a map won't use this constructor, and therefore never needs to call this method.
+         */
+        return defaultConstructorReference.updateAndGet(reference -> {
+            if (reference != null) {
+                return reference;
+            }
```

Comment:
If the class does not have any default ctor, then `ReflectionUtils.getDefaultConstructor` will return null, in which case we would use reflection to try and get the default ctor every time this method is called.
Instead, we need to distinguish between an actual `null` result and `getDefaultConstructor` not yet being invoked.

Reply:
Makes sense, I'll adjust the code.